src/tools/wasm-shell.cpp

```diff
@@ -41,300 +40,380 @@ Name ASSERT_INVALID("assert_invalid");
-  if (wasm) {
-    // prefix make_unique to work around visual studio bugs
-    auto tempInterface = wasm::make_unique<ShellExternalInterface>();
-    auto tempInstance =
-      wasm::make_unique<ModuleInstance>(*wasm, tempInterface.get());
-    interfaces[moduleName].swap(tempInterface);
-    instances[moduleName].swap(tempInstance);
-    instance = instances[moduleName].get();
-    if (entry.is()) {
-      Function* function = wasm->getFunction(entry);
-      if (!function) {
-        std::cerr << "Unknown entry " << entry << std::endl;
-      } else {
-        LiteralList arguments;
-        for (const auto& param : function->sig.params) {
-          arguments.push_back(Literal(param));
-        }
-        try {
-          instance->callExport(entry, arguments);
-        } catch (ExitException&) {
-        }
-      }
+  void parse(Element& s) {
+    IString id = s[0]->str();
+    if (id == MODULE) {
+      parseModule(s);
+    } else if (id == REGISTER) {
+      parseRegister(s);
+    } else if (id == INVOKE) {
+      parseOperation(s);
+    } else if (id == ASSERT_RETURN) {
+      parseAssertReturn(s);
+    } else if (id == ASSERT_TRAP) {
+      parseAssertTrap(s);
+    } else if ((id == ASSERT_INVALID) || (id == ASSERT_MALFORMED)) {
+      parseModuleAssertion(s);
     }
   }
-  while (*i < root->size()) {
-    Element& curr = *(*root)[*i];
-    IString id = curr[0]->str();
-    if (id == MODULE) {
-      break;
+
+  Module* parseModule(Element& s) {
+    if (options.debug) {
+      std::cerr << "parsing s-expressions to wasm...\n";
     }
-    *checked = true;
-    Colors::red(std::cerr);
-    std::cerr << *i << '/' << (root->size() - 1);
     Colors::green(std::cerr);
-    std::cerr << " CHECKING: ";
+    std::cerr << "BUILDING MODULE [line: " << s.line << "]\n";
     Colors::normal(std::cerr);
-    std::cerr << curr;
+    auto module = wasm::make_unique<Module>();
+    auto builder =
+      wasm::make_unique<SExpressionWasmBuilder>(*module, s, IRProfile::Normal);
+    auto moduleName = module->name;
+    lastModule = module->name;
+    builders[moduleName].swap(builder);
+    modules[moduleName].swap(module);
+    modules[moduleName]->features = FeatureSet::All;
+    bool valid = WasmValidator().validate(*modules[moduleName]);
+    if (!valid) {
+      std::cout << *modules[moduleName] << '\n';
+      Fatal() << "module failed to validate, see above";
+    }
+
+    instantiate(modules[moduleName].get());
+
+    return modules[moduleName].get();
+  }
+
+  void parseRegister(Element& s) {
+    auto instance = instances[lastModule];
+    if (!instance) {
+      Fatal() << "register called without a module";
+    }
+    auto name = s[1]->str();
+    linkedInstances[name] = instance;
+
+    // swap to the new name in all maps
+    modules[name].swap(modules[lastModule]);
+    builders[name].swap(builders[lastModule]);
+    interfaces[name].swap(interfaces[lastModule]);
+    instances[name].swap(instances[lastModule]);
```

Comment:
Shouldn't the items at `lastModule` be deleted, not just left as default values?

Reply:
You're right! Fixed now.